tests/test_data.py

```diff
@@ -0,0 +1,193 @@
+ACT_SPACES = SPACES
+
+
+def _check_1d_shape(fn: Callable[[np.ndarray], Any], length: float, expected_msg: str):
+    for shape in [(), (length, 1), (length, 2), (length - 1,), (length + 1,)]:
+        with pytest.raises(ValueError, match=expected_msg):
+            fn(np.zeros(shape))
+
+
+@pytest.mark.parametrize("obs_space", OBS_SPACES)
+@pytest.mark.parametrize("act_space", ACT_SPACES)
+@pytest.mark.parametrize("length", [0, 1, 2, 10])
+def test_valid_trajectories(
+    obs_space: gym.Space, act_space: gym.Space, length: int
+) -> None:
+    """Checks trajectories can be created for a variety of lengths and spaces."""
+    obs = np.array([obs_space.sample() for _ in range(length + 1)])
+    acts = np.array([act_space.sample() for _ in range(length)])
+    infos = [{} for _ in range(length)]
+    rews = np.random.randn(length)
+
+    traja = data.Trajectory(obs=obs, acts=acts, infos=None)
+    trajb = data.Trajectory(obs=obs, acts=acts, infos=infos)
+    trajc = data.TrajectoryWithRew(obs=obs, acts=acts, infos=None, rews=rews)
+    trajd = data.TrajectoryWithRew(obs=obs, acts=acts, infos=infos, rews=rews)
+    for traj in [traja, trajb, trajc, trajd]:
+        assert len(traj) == length
+
+
+@pytest.mark.parametrize("obs_space", OBS_SPACES)
+@pytest.mark.parametrize("act_space", ACT_SPACES)
+@pytest.mark.parametrize("length", [1, 2, 10])
+def test_invalid_trajectories(
+    obs_space: gym.Space, act_space: gym.Space, length: int
+) -> None:
+    """Checks input validation catches space and dtype related errors."""
+    obs = np.array([obs_space.sample() for _ in range(length + 1)])
+    acts = np.array([act_space.sample() for _ in range(length)])
+    infos = [{} for _ in range(length)]
+    rews = np.random.randn(length)
+
+    for cls in [data.Trajectory, functools.partial(data.TrajectoryWithRew, rews=rews)]:
+        with pytest.raises(
+            ValueError, match=r"expected one more observations than actions.*"
+        ):
+            cls(obs=obs[:-1], acts=acts, infos=None)
+        with pytest.raises(
+            ValueError, match=r"expected one more observations than actions.*"
+        ):
+            cls(obs=obs, acts=acts[:-1], infos=None)
+
+        with pytest.raises(
+            ValueError, match=r"infos when present must be present for each action.*"
+        ):
+            cls(obs=obs, acts=acts, infos=infos[:-1])
+        with pytest.raises(
+            ValueError, match=r"infos when present must be present for each action.*"
+        ):
+            cls(obs=obs[:-1], acts=acts[:-1], infos=infos)
+
+    _check_1d_shape(
+        fn=lambda rews: data.TrajectoryWithRew(
+            obs=obs, acts=acts, infos=infos, rews=rews
+        ),
+        length=length,
+        expected_msg=r"rewards must be 1D array.*",
+    )
+
+    with pytest.raises(ValueError, match=r"rewards dtype.* not a float"):
+        data.TrajectoryWithRew(
+            obs=obs, acts=acts, infos=infos, rews=np.zeros(length, dtype=np.int)
+        )
```

Comment:
These assertions seem like a good candidates for refactoring with `dataclasses.replace`, since you are only replacing one field and repeating three.